File: sample/java/11_lissajous_particles.java
public final class _11_lissajous_particles {
    private _11_lissajous_particles() {
    }


    // 11: Sample that outputs Lissajous-motion particles as a GIF.

    public static java.util.ArrayList<Long> color_palette() {
        java.util.ArrayList<Long> p = new java.util.ArrayList<Long>();
        for (long i = 0L; i < 256L; i += 1L) {
            long r = i;
            long g = i * 3L % 256L;
            long b = 255L - i;
            p.add(r);
            p.add(g);
            p.add(b);
        }
        return new java.util.ArrayList<Long>(p);
    }

    public static void run_11_lissajous_particles() {
        long w = 320L;
        long h = 240L;
        long frames_n = 360L;
        long particles = 48L;
        String out_path = "sample/out/11_lissajous_particles.gif";
        double start = (System.nanoTime() / 1000000000.0);
        java.util.ArrayList<java.util.ArrayList<Long>> frames = new java.util.ArrayList<java.util.ArrayList<Long>>();
        for (long t = 0L; t < frames_n; t += 1L) {
            java.util.ArrayList<Long> frame = PyRuntime.__pytra_bytearray(w * h);
            double __hoisted_cast_1 = ((double)(t));
            for (long p = 0L; p < particles; p += 1L) {
                double phase = ((double)(p)) * 0.261799;
                long x = PyRuntime.__pytra_int(((double)(w)) * 0.5 + ((double)(w)) * 0.38 * Math.sin(0.11 * __hoisted_cast_1 + phase * 2.0));
                long y = PyRuntime.__pytra_int(((double)(h)) * 0.5 + ((double)(h)) * 0.38 * Math.sin(0.17 * __hoisted_cast_1 + phase * 3.0));
                long color = 30L + p * 9L % 220L;
                for (long dy = (-2L); dy < 3L; dy += 1L) {
                    for (long dx = (-2L); dx < 3L; dx += 1L) {
                        long xx = x + dx;
                        long yy = y + dy;
                        if (((xx >= 0L) && (xx < w) && (yy >= 0L) && (yy < h))) {
                            long d2 = dx * dx + dy * dy;
                            if ((d2 <= 4L)) {
                                long idx = yy * w + xx;
                                long v = color - d2 * 20L;
                                v = Math.max(0L, v);
                                if ((v > ((Long)(frame.get((int)((((idx) < 0L) ? (((long)(frame.size())) + (idx)) : (idx)))))))) {
                                    frame.set((int)((((idx) < 0L) ? (((long)(frame.size())) + (idx)) : (idx))), v);
                                }
                            }
                        }
                    }
                }
            }
            frames.add(new java.util.ArrayList<Long>(frame));
        }
        PyRuntime.__pytra_noop(out_path, w, h, frames, color_palette());
        double elapsed = (System.nanoTime() / 1000000000.0) - start;
        System.out.println(String.valueOf("output:") + " " + String.valueOf(out_path));
        System.out.println(String.valueOf("frames:") + " " + String.valueOf(frames_n));
        System.out.println(String.valueOf("elapsed_sec:") + " " + String.valueOf(elapsed));
    }

    public static void main(String[] args) {
        run_11_lissajous_particles();
    }
}

File: sample/lua/11_lissajous_particles.lua
dofile((debug.getinfo(1, "S").source:sub(2):match("^(.*[\\/])") or "") .. "py_runtime.lua")

-- from __future__ import annotations as annotations (not yet mapped)
local math = __pytra_math_module()
local perf_counter = __pytra_perf_counter
local save_gif = __pytra_save_gif

-- 11: Sample that outputs Lissajous-motion particles as a GIF.

function color_palette()
    p = __pytra_bytearray()
    for i = 0, 256 - 1 do
        r = i
        g = ((i * 3) % 256)
        b = (255 - i)
        table.move({r, g, b}, 1, 3, #(p) + 1, p)
    end
    return __pytra_bytes(p)
end

function run_11_lissajous_particles()
    w = 320
    h = 240
    frames_n = 360
    particles = 48
    out_path = "sample/out/11_lissajous_particles.gif"
    
    start = perf_counter()
    local frames = {  }
    
    for t = 0, frames_n - 1 do
        frame = __pytra_bytearray((w * h))
        local __hoisted_cast_1 = __pytra_float(t)
        
        for p = 0, particles - 1 do
            phase = (p * 0.261799)
            x = __pytra_int(((w * 0.5) + ((w * 0.38) * math.sin(((0.11 * __hoisted_cast_1) + (phase * 2.0))))))
            y = __pytra_int(((h * 0.5) + ((h * 0.38) * math.sin(((0.17 * __hoisted_cast_1) + (phase * 3.0))))))
            color = (30 + ((p * 9) % 220))
            
            for dy = (-2), 3 - 1 do
                for dx = (-2), 3 - 1 do
                    xx = (x + dx)
                    yy = (y + dy)
                    if ((xx >= 0) and (xx < w) and (yy >= 0) and (yy < h)) then
                        d2 = ((dx * dx) + (dy * dy))
                        if (d2 <= 4) then
                            idx = ((yy * w) + xx)
                            v = (color - (d2 * 20))
                            v = math.max(0, v)
                            if (v > frame[(((idx) < 0) and (#(frame) + (idx) + 1) or ((idx) + 1))]) then
                                frame[(((idx) < 0) and (#(frame) + (idx) + 1) or ((idx) + 1))] = v
                            end
                        end
                    end
                end
            end
        end
        table.insert(frames, __pytra_bytes(frame))
    end
    save_gif(out_path, w, h, frames, color_palette())
    elapsed = (perf_counter() - start)
    __pytra_print("output:", out_path)
    __pytra_print("frames:", frames_n)
    __pytra_print("elapsed_sec:", elapsed)
end


run_11_lissajous_particles()
